### PR TITLE
Clean up data loader arguments, tests, and add docstrings

### PR DESCRIPTION
**Pull request by lostella**
*Description of changes:* This is mostly adding docstrings for the data loading tooling, but I took the chance to also clean up a few things, and rewrite the data loading tests.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/gluonts/dataset/loader.py [hunk 1]
```diff
@@ -240,14 +312,41 @@ def ValidationDataLoader(
...
+    Returns
+    -------
+    Iterable[DataBatch]
+        An iterable sequence of batches.
+    """
     return DataLoader(
         data_iterable=TransformedDataset(dataset, transform, is_train=True),
         batch_size=batch_size,
         stack_fn=stack_fn,
-        decode_fn=decode_fn,
```
[the remark this answers is not in the document]
Comment (lostella) [reply src/gluonts/dataset/loader.py:250]: whoops, good catch, I forgot to put these back

File: src/gluonts/dataset/loader.py [hunk 2]
```diff
@@ -168,6 +193,9 @@ def __init__(
         num_prefetch: Optional[int] = None,
         decode_fn: Callable = lambda x: x,
     ) -> None:
+        assert num_workers is None or num_workers > 0
```
Comment (jaheba) [inline src/gluonts/dataset/loader.py:196]: There is `core.ty.checked`, which we could also use for these assertions.

File: src/gluonts/dataset/loader.py [hunk 3]
```diff
@@ -158,6 +158,31 @@ def win32_guard(num_workers: Optional[int]) -> Optional[int]:
 
 
 class DataLoader(Iterable[DataBatch]):
+    """Iterates a datasets and puts its entries into batches of a given size.
```
Comment (jaheba) [inline src/gluonts/dataset/loader.py:161]: To quote PEP 257:

>  It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...".

So this should be "Iterate a dataset and put ...`

File: src/gluonts/dataset/loader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaheba) [reply src/gluonts/dataset/loader.py:250]: We don't use `decode` in validation?

[the remark this answers is not in the document]
Comment (lostella) [reply src/gluonts/dataset/loader.py:250]: No, exactly: that's needed when using multiple processes only, and that's not currently supported in this case.  I tried making it work, but I had issues with tests on Linux. Note that multiple workers were not supported here even prior to this PR, only this `decode_fn` argument was left there for some reason (but had no effect)

File: src/gluonts/dataset/loader.py (hunk 2 above)
Comment (lostella) [reply src/gluonts/dataset/loader.py:196]: should I use pydantic types to enforce positiveness?

**Review by jaheba — APPROVED** [review]

**Pull request merged**